File: src/main/java/org/simbrain/network/gui/NetworkActionManager.java
///*
// * Part of Simbrain--a java-based neural network kit
// * Copyright (C) 2005,2007 The Authors.  See http://www.simbrain.net/credits
// *
// * This program is free software; you can redistribute it and/or modify
// * it under the terms of the GNU General Public License as published by
// * the Free Software Foundation; either version 2 of the License, or
// * (at your option) any later version.
// *
// * This program is distributed in the hope that it will be useful,
// * but WITHOUT ANY WARRANTY; without even the implied warranty of
// * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// * GNU General Public License for more details.
// *
// * You should have received a copy of the GNU General Public License
// * along with this program; if not, write to the Free Software
// * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
// */
//package org.simbrain.network.gui;
//
//import org.simbrain.network.connections.*;
//import org.simbrain.network.groups.NeuronCollection;
//import org.simbrain.network.groups.NeuronGroup;
//import org.simbrain.network.groups.SynapseGroup;
//import org.simbrain.network.gui.actions.*;
//import org.simbrain.network.gui.actions.connection.ApplyConnectionAction;
//import org.simbrain.network.gui.actions.connection.ClearSourceNeurons;
//import org.simbrain.network.gui.actions.connection.SetSourceNeurons;
//import org.simbrain.network.gui.actions.edit.*;
//import org.simbrain.network.gui.actions.modelgroups.AddGroupAction;
//import org.simbrain.network.gui.actions.modelgroups.NeuronCollectionAction;
//import org.simbrain.network.gui.actions.network.*;
//import org.simbrain.network.gui.actions.neuron.NewNeuronAction;
//import org.simbrain.network.gui.actions.neuron.SetNeuronPropertiesAction;
//import org.simbrain.network.gui.actions.neuron.ShowPrioritiesAction;
//import org.simbrain.network.gui.actions.selection.*;
//import org.simbrain.network.gui.actions.synapse.*;
//import org.simbrain.network.gui.actions.toolbar.ShowEditToolBarAction;
//import org.simbrain.network.gui.actions.toolbar.ShowMainToolBarAction;
//import org.simbrain.network.gui.actions.toolbar.ShowRunToolBarAction;
//import org.simbrain.network.gui.dialogs.group.NeuronGroupDialog;
//import org.simbrain.network.gui.dialogs.network.*;
//
//import javax.swing.*;
//import java.util.*;
//
///**
// * Contains instances of actions, some of them which maintain state, for common
// * reference across Simbrain. They are accessed using the name of the action class
// * or a related class. Simply inspect the code if not sure.
// */
//public final class NetworkActionManager {
//
//    /**
//     * Reference to NetworkPanel.
//     */
//    private final NetworkPanel networkPanel;
//
//    /**
//     * Map storing class -> action associations
//     */
//    private transient Map<Class, AbstractAction> actions = new HashMap<>();
//
//    /**
//     * Create a new network action manager for the specified network panel.
//     *
//     * @param networkPanel networkPanel, must not be null
//     */
//    NetworkActionManager(final NetworkPanel networkPanel) {
//
//        this.networkPanel = networkPanel;
//
//        if (networkPanel == null) {
//            throw new IllegalArgumentException("networkPanel must not be null");
//        }
//
//        // Could have used reflection but it would only make things simpler on this page and the code below is easy to
//        // follow, and this class is easy to use externally
//
//        actions.put(ZoomToFitPageAction.class, new ZoomToFitPageAction(networkPanel));
//        actions.put(NewNeuronAction.class, new NewNeuronAction(networkPanel));
//        actions.put(ClearNodeActivationsAction.class, new ClearNodeActivationsAction(networkPanel));
//        actions.put(RandomizeObjectsAction.class, new RandomizeObjectsAction(networkPanel));
//        actions.put(SelectAllAction.class, new SelectAllAction(networkPanel));
//        actions.put(DeleteAction.class, new DeleteAction(networkPanel));
//        actions.put(CopyAction.class, new CopyAction(networkPanel));
//        actions.put(CutAction.class, new CutAction(networkPanel));
//        actions.put(PasteAction.class, new PasteAction(networkPanel));
//        actions.put(IterateNetworkAction.class, new IterateNetworkAction(networkPanel));
//        actions.put(RunNetworkAction.class, new RunNetworkAction(networkPanel));
//        actions.put(StopNetworkAction.class, new StopNetworkAction(networkPanel));
//        actions.put(ShowDebugAction.class, new ShowDebugAction(networkPanel));
//        actions.put(ShowNetworkPreferencesAction.class, new ShowNetworkPreferencesAction(networkPanel));
//        actions.put(AlignHorizontalAction.class, new AlignHorizontalAction(networkPanel));
//        actions.put(AlignVerticalAction.class, new AlignVerticalAction(networkPanel));
//        actions.put(SpaceVerticalAction.class, new SpaceVerticalAction(networkPanel));
//        actions.put(SpaceHorizontalAction.class, new SpaceHorizontalAction(networkPanel));
//        actions.put(SetNeuronPropertiesAction.class, new SetNeuronPropertiesAction(networkPanel));
//        actions.put(SetSynapsePropertiesAction.class, new SetSynapsePropertiesAction(networkPanel));
//        actions.put(SelectAllWeightsAction.class, new SelectAllWeightsAction(networkPanel));
//        actions.put(SelectAllNeuronsAction.class, new SelectAllNeuronsAction(networkPanel));
//        actions.put(ShowMainToolBarAction.class, new ShowMainToolBarAction(networkPanel));
//        actions.put(ShowEditToolBarAction.class, new ShowEditToolBarAction(networkPanel));
//        actions.put(ShowRunToolBarAction.class, new ShowRunToolBarAction(networkPanel));
//        actions.put(SelectIncomingWeightsAction.class, new SelectIncomingWeightsAction(networkPanel));
//        actions.put(SelectOutgoingWeightsAction.class, new SelectOutgoingWeightsAction(networkPanel));
//        actions.put(SetTextPropertiesAction.class, new SetTextPropertiesAction(networkPanel));
//        actions.put(ShowWeightMatrixAction.class, new ShowWeightMatrixAction(networkPanel));
//        actions.put(ShowAdjustSynapsesDialog.class, new ShowAdjustSynapsesDialog(networkPanel));
//        actions.put(ShowAdjustConnectivityDialog.class, new ShowAdjustConnectivityDialog(networkPanel));
//        actions.put(ShowNetworkUpdaterDialog.class, new ShowNetworkUpdaterDialog(networkPanel));
//        actions.put(TestInputAction.class, new TestInputAction(networkPanel));
//        actions.put(ShowLayoutDialogAction.class, new ShowLayoutDialogAction(networkPanel));
//        actions.put(SetSourceNeurons.class, new SetSourceNeurons(networkPanel));
//        actions.put(ClearSourceNeurons.class, new ClearSourceNeurons(networkPanel));
//        actions.put(ShowPrioritiesAction.class, new ShowPrioritiesAction(networkPanel));
//        actions.put(ShowWeightsAction.class, new ShowWeightsAction(networkPanel));
//        actions.put(SelectionEditModeAction.class, new SelectionEditModeAction(networkPanel));
//        actions.put(TextEditModeAction.class, new TextEditModeAction(networkPanel));
//        actions.put(WandEditModeAction.class, new WandEditModeAction(networkPanel));
//
//        actions.put(AllToAll.class, new ApplyConnectionAction(networkPanel, new AllToAll(), "All to all"));
//        actions.put(OneToOne.class, new ApplyConnectionAction(networkPanel, new OneToOne(), "One-to-one"));
//        actions.put(RadialGaussian.class, new ApplyConnectionAction(networkPanel, new RadialGaussian(), "Radial (Gaussian)"));
//        actions.put(RadialSimple.class, new ApplyConnectionAction(networkPanel, new RadialSimple(), "Radial (Simple)"));
//        actions.put(Sparse.class, new ApplyConnectionAction(networkPanel, new Sparse(), "Sparse"));
//
//        actions.put(NeuronGroup.class, new AddGroupAction(networkPanel,
//                NeuronGroupDialog.class, "Add Neuron Group"));
//        actions.put(AddSynapseGroupAction.class, new AddSynapseGroupAction(networkPanel));
//        actions.put(NeuronCollectionAction.class, new NeuronCollectionAction(networkPanel));
//
//
//    }
//
//    public List<Action> getNetworkModeActions() {
//        return Arrays.asList(new Action[]{actions.get(SelectionEditModeAction.class), actions.get(TextEditModeAction.class),
//                actions.get(WandEditModeAction.class)});
//    }
//
//    public List<Action> getNetworkControlActions() {
//        return Arrays.asList(new Action[]{actions.get(RunNetworkAction.class), actions.get(StopNetworkAction.class)});
//    }
//
//    public List<Action> getClipboardActions() {
//        return Arrays.asList(new Action[]{actions.get(CopyAction.class), actions.get(CutAction.class),
//                actions.get(PasteAction.class)});
//    }
//
//    public List<Action> getNetworkEditingActions() {
//        return Arrays.asList(new Action[]{actions.get(NewNeuronAction.class), actions.get(DeleteAction.class)});
//    }
//
//    public JMenu getConnectionMenu() {
//        JMenu cm = new JMenu("Connect Neurons");
//        cm.add(getMenuItem(AllToAll.class));
//        cm.add(getMenuItem(OneToOne.class));
//        cm.add(getMenuItem(RadialGaussian.class));
//        cm.add(getMenuItem(RadialSimple.class));
//        cm.add(getMenuItem(Sparse.class));
//        return cm;
//    }
//
//    public List<Action> getNewNetworkActions() {
//        return Arrays.asList(new Action[]{new AddGroupAction(networkPanel, BackpropCreationDialog.class, "Backprop"),
//                new AddGroupAction(networkPanel, CompetitiveCreationDialog.class, "Competitive Network"),
//                new AddGroupAction(networkPanel, FeedForwardCreationDialog.class, "Feed Forward Network"),
//                new AddGroupAction(networkPanel, HopfieldCreationDialog.class, "Hopfield"),
//                new AddGroupAction(networkPanel, LMSCreationDialog.class, "LMS (Least Mean Squares)"),
//                new AddGroupAction(networkPanel, SOMCreationDialog.class, "SOM Network"),
//                new AddGroupAction(networkPanel, SRNCreationDialog.class, "SRN (Simple Recurrent Network)")});
//    }
//
//    public JMenu getNewNetworkMenu() {
//        JMenu ret = new JMenu("Insert Network");
//        for (Action action : getNewNetworkActions()) {
//            ret.add(action);
//        }
//        return ret;
//    }
//
//    /**
//     * Returns on action based on its associated name (see above)
//     */
//    public Action getAction(Class clazz) {
//        return actions.get(clazz);
//    }
//
//    /**
//     * Gets an action within a {@link JCheckBoxMenuItem}.  Can just be
//     * used as a menu item without using the checkbox.
//     */
//    public JCheckBoxMenuItem getMenuItem(Class clazz) {
//        return new JCheckBoxMenuItem(getAction(clazz));
//    }
//
//    /**
//     * Gets an action within a {@link JCheckBoxMenuItem}.
//     *
//     * @param clazz    class associated with action
//     * @param checked whether it should initially be checked.
//     */
//    public JCheckBoxMenuItem getMenuItem(Class clazz, boolean checked) {
//        JCheckBoxMenuItem menuItem = new JCheckBoxMenuItem(getAction(clazz));
//        menuItem.setSelected(checked);
//        return menuItem;
//    }
//
//}

File: src/main/kotlin/org/simbrain/network/gui/NetworkActionManager.kt
package org.simbrain.network.gui

import org.simbrain.network.gui.actions.SetTextPropertiesAction
import org.simbrain.network.gui.actions.ShowDebugAction
import org.simbrain.network.gui.actions.ShowLayoutDialogAction
import org.simbrain.network.gui.actions.TestInputAction
import org.simbrain.network.gui.actions.connection.ClearSourceNeurons
import org.simbrain.network.gui.actions.connection.SetSourceNeurons
import org.simbrain.network.gui.actions.edit.*
import org.simbrain.network.gui.actions.modelgroups.AddGroupAction
import org.simbrain.network.gui.actions.network.*
import org.simbrain.network.gui.actions.neuron.NewNeuronAction
import org.simbrain.network.gui.actions.neuron.SetNeuronPropertiesAction
import org.simbrain.network.gui.actions.neuron.ShowPrioritiesAction
import org.simbrain.network.gui.actions.selection.*
import org.simbrain.network.gui.actions.synapse.*
import org.simbrain.network.gui.actions.toolbar.ShowEditToolBarAction
import org.simbrain.network.gui.actions.toolbar.ShowMainToolBarAction
import org.simbrain.network.gui.actions.toolbar.ShowRunToolBarAction
import org.simbrain.network.gui.dialogs.group.NeuronGroupDialog
import org.simbrain.network.gui.dialogs.network.*
import javax.swing.AbstractAction
import javax.swing.JCheckBoxMenuItem
import javax.swing.JMenu


class NetworkActionManager(val networkPanel: NetworkPanel) {

    val actions = mapOf(
            ZoomToFitPageAction::class.java to ZoomToFitPageAction(networkPanel),
            NewNeuronAction::class.java to NewNeuronAction(networkPanel),
            ClearNodeActivationsAction::class.java to ClearNodeActivationsAction(networkPanel),
            RandomizeObjectsAction::class.java to RandomizeObjectsAction(networkPanel),
            SelectAllAction::class.java to SelectAllAction(networkPanel),
            DeleteAction::class.java to DeleteAction(networkPanel),
            CopyAction::class.java to CopyAction(networkPanel),
            CutAction::class.java to CutAction(networkPanel),
            PasteAction::class.java to PasteAction(networkPanel),
            IterateNetworkAction::class.java to IterateNetworkAction(networkPanel),
            RunNetworkAction::class.java to RunNetworkAction(networkPanel),
            StopNetworkAction::class.java to StopNetworkAction(networkPanel),
            ShowDebugAction::class.java to ShowDebugAction(networkPanel),
            ShowNetworkPreferencesAction::class.java to ShowNetworkPreferencesAction(networkPanel),
            AlignHorizontalAction::class.java to AlignHorizontalAction(networkPanel),
            AlignVerticalAction::class.java to AlignVerticalAction(networkPanel),
            SpaceVerticalAction::class.java to SpaceVerticalAction(networkPanel),
            SpaceHorizontalAction::class.java to SpaceHorizontalAction(networkPanel),
            SetNeuronPropertiesAction::class.java to SetNeuronPropertiesAction(networkPanel),
            SetSynapsePropertiesAction::class.java to SetSynapsePropertiesAction(networkPanel),
            SelectAllWeightsAction::class.java to SelectAllWeightsAction(networkPanel),
            SelectAllNeuronsAction::class.java to SelectAllNeuronsAction(networkPanel),
            ShowMainToolBarAction::class.java to ShowMainToolBarAction(networkPanel),
            ShowEditToolBarAction::class.java to ShowEditToolBarAction(networkPanel),
            ShowRunToolBarAction::class.java to ShowRunToolBarAction(networkPanel),
            SelectIncomingWeightsAction::class.java to SelectIncomingWeightsAction(networkPanel),
            SelectOutgoingWeightsAction::class.java to SelectOutgoingWeightsAction(networkPanel),
            SetTextPropertiesAction::class.java to SetTextPropertiesAction(networkPanel),
            ShowWeightMatrixAction::class.java to ShowWeightMatrixAction(networkPanel),
            ShowAdjustSynapsesDialog::class.java to ShowAdjustSynapsesDialog(networkPanel),
            ShowAdjustConnectivityDialog::class.java to ShowAdjustConnectivityDialog(networkPanel),
            ShowNetworkUpdaterDialog::class.java to ShowNetworkUpdaterDialog(networkPanel),
            TestInputAction::class.java to TestInputAction(networkPanel),
            ShowLayoutDialogAction::class.java to ShowLayoutDialogAction(networkPanel),
            SetSourceNeurons::class.java to SetSourceNeurons(networkPanel),
            ClearSourceNeurons::class.java to ClearSourceNeurons(networkPanel),
            ShowPrioritiesAction::class.java to ShowPrioritiesAction(networkPanel),
            ShowWeightsAction::class.java to ShowWeightsAction(networkPanel),
            SelectionEditModeAction::class.java to SelectionEditModeAction(networkPanel),
            TextEditModeAction::class.java to TextEditModeAction(networkPanel),
            WandEditModeAction::class.java to WandEditModeAction(networkPanel)
    )

    inline fun <reified T : AbstractAction> getAction() = actions[T::class.java]

    //TODO: Yulin please check these.  Could use reified below?
    /**
     * Gets an action within a [JCheckBoxMenuItem].  Can just be
     * used as a menu item without using the checkbox.
     */
     inline fun <reified T : AbstractAction> getMenuItem(): JCheckBoxMenuItem {
        return JCheckBoxMenuItem(this.getAction<T>());
    }
    inline fun <reified T : AbstractAction> getMenuItem(checked:Boolean): JCheckBoxMenuItem {
        val menuItem = getMenuItem<T>()
        menuItem.setSelected(checked);
        return menuItem;
    }
    val clipboardActions
        get() = listOf(getAction<CopyAction>(),getAction<CutAction>(),getAction<PasteAction>())

    val networkControlActions
        get() = listOf(getAction<RunNetworkAction>(),getAction<StopNetworkAction>())

    val networkEditingActions
        get() = listOf(getAction<NewNeuronAction>(),getAction<DeleteAction>())

    // TODO: Yulin, isn't the below more verbose than the above?
    val networkModeActions
        get() = listOf<Class<out AbstractAction>>(
                SelectionEditModeAction::class.java,
                TextEditModeAction::class.java,
                WandEditModeAction::class.java
        ).map { actions[it]!! }

    val neuronGroupAction = AddGroupAction(networkPanel, NeuronGroupDialog::class.java, "Add Neuron Group")

    private fun AbstractAction.toMenuItem() = JCheckBoxMenuItem(this)

    val newNetworkActions get() = listOf(
            AddGroupAction(networkPanel, BackpropCreationDialog::class.java, "Backprop"),
            AddGroupAction(networkPanel, CompetitiveCreationDialog::class.java, "Competitive Network"),
            AddGroupAction(networkPanel, FeedForwardCreationDialog::class.java, "Feed Forward Network"),
            AddGroupAction(networkPanel, HopfieldCreationDialog::class.java, "Hopfield"),
            AddGroupAction(networkPanel, LMSCreationDialog::class.java, "LMS (Least Mean Squares)"),
            AddGroupAction(networkPanel, SOMCreationDialog::class.java, "SOM Network"),
            AddGroupAction(networkPanel, SRNCreationDialog::class.java, "SRN (Simple Recurrent Network)")
    )

    val newNetworkMenu get() = JMenu("Insert Network").apply {
        newNetworkActions.forEach { add(it) }
    }

}